### ISSUE TITLE
Dragged element duplication

### ISSUE THREAD

**Issue by mgthomas99** [hejianxian/vddl#4]
I would like to report a bug.

**VDDL Version**: 0.5.2
**Vue Version**: 2.3.3

When I drag-and-drop `<vddl-draggable>` elements within a `<vddl-list>` too quickly while there is text selected on the webpage, the dragged element sometimes clones itself and replaces a different element in the list.

I have created an example project [here](https://github.com/mgthomas99/__vddl-duplication-bug-repro), although the bug can occur in any project.

How to reproduce:
1. Create a `<vddl-list>` and fill it with an arbitrary number of `<vddl-draggable>` items.
2. Press Ctrl+A to select all text on the page. (I have noticed that the bug is more likely to happen if there is a lot of text selected).
2. Repeatedly drag-and-drop the same item from the list as quickly as possible. It will eventually clone itself by replacing another element in the list (typically the element closest to where you dropped the item).

**Comment by baixiyang** [comment]
how to resolve this problem

**Comment by baixiyang** [comment]
我从上往下拖不会有问题，但是从下往上拖就会有问题。而且这个问题有时候会发生有时候突然又正常了。。。。

**Comment by hejianxian** [comment]
@baixiyang 谢谢关注！

我观察过其他人写的基于H5的拖动库，大部分都不会出现这种问题，但是我暂时还没找到解决方法。而且最近一直在加班，没太多时间去处理这个问题，实在抱歉。

如果你有任何解决思路，请告诉我一下，谢谢！

**Comment by AWulkan** [comment]
My boss was able to find a way to recreate the bug (Please don't mind the awful design, it's only temporary).
![bug](https://imgur.com/933e220c-c78e-4b59-b49b-7c98cd4b3b56)

I haven't tried recreating it like this outside of our project yet though.

**Comment by AWulkan** [comment]
Hey again. I decided to try it out with your example code for VDDL nested lists, and I got some weird results...

In this gif I first manage to recreate the bug in my own local dev environment, running your example code (I haven't changed anything). But when I switch over to your website and do the same, it works.
https://imgur.com/a/DPvcX

I have tried this over and over many times now and about 80% of the time I get the bug in the dev environment, but never on your website.

**Comment by AWulkan** [comment]
I have made some progress.

I believe it's this line that is removing the wrong item from the list:
https://github.com/hejianxian/vddl/blob/5ffb105ae49ba1fce59b9dd49161c7050d56bcc4/src/components/vddl-draggable.vue#L87

I did some logging like this: https://imgur.com/a/2j9KW

If I start with a list like this:
```js
0: {type: "item", id: "1"}
1: {type: "item", id: "2"}
2: {type: "item", id: "3"}
```

And then drag item 3 up between item 1 and 2, I get this logged to the console:
```js
wrapper.splice: 2

wrapper before:
0: {type: "item", id: "1"}
1: {type: "item", id: "3"}
2: {type: "item", id: "2"}
3: {type: "item", id: "3"}

wrapper after:
0: {type: "item", id: "1"}
1: {type: "item", id: "3"}
2: {type: "item", id: "3"}
```

This indicates that we are splicing index 2 from the list, which is `{type: "item", id: "2"}`.

If I change the code to: `this.wrapper.splice(this.index + 1, 1);` I can now drag upwards without problem!
However, now the bug appears when I drag downwards instead.

@hejianxian What do you think? Got any ideas?

**Comment by gjactat** [comment]
Hello,
The bug seems to be related to the version of Vue. I've built a small project on CodeSandbox with the exact code of your "horizontal" demo (https://codesandbox.io/s/j387n4j3y3). Note that CodeSandbox makes it easy to change the version of packages dependencies, thus comparing the ok version with the ko version.
The bug occurs with a dependency of Vue 2.5.6 but doesn't occur with Vue 2.4.2 (which is the version of your Gihtub package.json project).

This doesn't solve the problem but it gives us a hint. Something has probably changed since version 2.4.2 that broke your component logic. Maybe related to reactivity or handlers execution order ?

**Comment by AWulkan** [comment]
@gjactat I tested it a bit and it appears that 2.5.0 is the breakpoint for when things started go wrong.

**Comment by gjactat** [comment]
Yes... Version 2.5.0 brings major changes !

**Comment by AWulkan** [comment]
Also, if you look at my comment I explain where the wrong element is being removed from the list. Maybe I can install Vue 4.2.2 and log the same stuff to see the difference.

**Comment by gjactat** [comment]
If you can downgrade to Vue 2.4.4, you can fix your issue.
I personnaly can't downgrade :-/

**Comment by AWulkan** [comment]
I am just running the example project that you can find in this repo when I'm testing now.

**Comment by gjactat** [comment]
I need to kick a new sandbox with vddl sources instead of package dependency to investigate deeper ( breakpoints and traces....). We're getting closer !

**Comment by AWulkan** [comment]
Ok, so I have installed the older version now and I can see that the right index is set to `this.index`.

Starting list:
```js
0: {type: "item", id: "1"}
1: {type: "item", id: "2"}
2: {type: "item", id: "3"}
```

And then drag item 3 up between item 1 and 2, I get this logged to the console:
```js
wrapper.splice: 3

wrapper before:
0: {type: "item", id: "1"}
1: {type: "item", id: "3"}
2: {type: "item", id: "2"}
3: {type: "item", id: "3"}

wrapper after:
0: {type: "item", id: "1"}
1: {type: "item", id: "3"}
2: {type: "item", id: "2"}
```

Notice how `this.index` is 3 this time.

**Comment by gjactat** [comment]
I've got a patch on my fork. This is a quick fix that relies on ids instead of indexes. I won't make a pull request as long as it changes the behaviour (vddl originally doesn't need ids, only for Vue v-for directive).
Anyway, it works for cases where your items have ids...

https://github.com/gjactat/vddl/commit/1102b2108230a0c743f8a2c74e477cd17a785942#diff-dfffb20e5ca8ccf0fe86081ed87328ea

**Comment by AWulkan** [comment]
I noticed that the Vuex example still suffers from this problem as well.
http://hejx.space/vddl-demo/#/vuex

Since it adds and deletes items through actions it wasn't solved by my pull request, but the solution should probably be similar now that we know the causation.

**Comment by baixiyang** [comment]
i think the reason is vue has not update the index after drop(when drop up ,one item insert before.vue should update the index to index +1,but not)
this.$nextTick(function () {
       this$1.wrapper.splice(this$1.wrapper.indexOf(this.draggable), 1);
 });
so just find the true index instead of this.index

**Comment by baixiyang** [comment]
this way cannot suit string and number list @hejianxian ，we need  stricter way to get the true index。

**Comment by AWulkan** [comment]
Yes, the index is passed before the list is updated for the Vuex example, we need a way to make sure that the real index is accessible.

**Comment by hejianxian** [comment]
@AWulkan I fixed it.

But I think the more serious problem is: press `Ctrl+A` to select all text on the page, and then, how to stop the text from being dragged while dragging.

**Comment by AWulkan** [comment]
@hejianxian Are you talking about the nested list when it's disabled? Well, I know that you can add `user-select: none;` in the css on the vddl-list to prevent selection of text. It worked for me in the project I'm working on.

**Comment by hejianxian** [comment]
😭 `Ctrl+A` can select all text on the page.

<img src="https://user-images.githubusercontent.com/6087438/33051691-3a6fe756-cea6-11e7-9ec5-e51f0137ddac.png" width="300" />

when dragging element:

<img src="https://user-images.githubusercontent.com/6087438/33051790-a9c8ee40-cea6-11e7-982a-6d689b4a940b.png" width="300" />

🐽 This is a big bug.

**Comment by baixiyang** [comment]
duplication element is the biggest bug i think. it will happen certainly,  when drag up ward. if the bug exit , no one will use this component at all。

**Comment by AWulkan** [comment]
@baixiyang The duplication bug should be solved. Where are you experiencing it? Have you updated to the latest version?

**Comment by baixiyang** [comment]
change line 'this.splice(this.index,1)' to 'this.splice(this.wrapper.indexOf(this.draggable), 1)'  is useful in my environment. this is just an adivse.

**Comment by arnoson** [comment]
I had a case where I had to access the final index (without the duplicated item in the list) inside the drop callback of a vddl-list and have a little workaround
```javascript
handleDrop(data) {
	let { index, list, item } = data
	list.splice(index, 0, item)

	const hasDuplicates = list.filter(el => el.id === item.id).length === 2
	const oldIndex = list.findIndex(el => el.id === item.id)

	if (hasDuplicates && oldIndex < index)
		index -= 1
	}

	// do something with the final index
},	
```

**Comment by huangqiang8438** [comment]
出现这个的问题是判断拖拽是缺少一个判断，判断当前拖拽的是向上移动还是向下移动
